File: client/src/model/PythonClass.ts
import PythonFunction from "./PythonFunction";

export default class PythonClass {

    readonly name: string;
    readonly decorators: string[];
    readonly superclasses: string[];
    readonly methods: PythonFunction[];
    readonly summary: string;
    readonly description: string;
    readonly fullDocstring: string;

    constructor(
        name: string,
        decorators: string[] = [],
        superclasses: string[] = [],
        methods: PythonFunction[] = [],
        summary: string = "",
        description: string = "",
        fullDocstring: string = "",
    ) {
        this.name = name;
        this.decorators = decorators;
        this.superclasses = superclasses;
        this.methods = methods;
        this.summary = summary;
        this.description = description;
        this.fullDocstring = fullDocstring;
    }

    toString() {
        let result = ""

        if (this.decorators.length > 0) {
            result += this.decorators.map(it => `@${it}`).join(" ")
            result += " "
        }

        result += `class ${this.name}`

        if (this.superclasses.length > 0) {
            result += `(${this.superclasses.join(", ")})`
        }

        return result
    }
}

File: api-editor/desktop/src/main/kotlin/com/larsreimann/apiEditor/features/ast/model/PythonClass.kt
package com.larsreimann.apiEditor.features.ast.model

import kotlinx.serialization.Serializable

interface PythonClass : PythonDeclaration {
    override val id: PythonClassId
    val decorators: List<PythonDecorator>
    val superclasses: List<PythonClassId>
    val documentation: PythonClassDocumentation
    val methods: List<PythonFunction>

    override fun children() = sequence {
        yieldAll(methods)
    }

    fun toMutablePythonClass() = MutablePythonClass(
        id = id,
        name = name,
        decorators = decorators,
        superclasses = superclasses,
        documentation = documentation,
        methods = methods.map { it.toMutablePythonFunction() },
    )
}

class MutablePythonClass(
    override val id: PythonClassId = PythonClassId(""),
    override var name: String = "",
    decorators: List<PythonDecorator> = emptyList(),
    superclasses: List<PythonClassId> = emptyList(),
    override var documentation: PythonClassDocumentation = PythonClassDocumentation(),
    methods: List<MutablePythonFunction> = listOf(),
) : MutablePythonDeclaration(), PythonClass {

    override val decorators = decorators.toMutableList()
    override val superclasses = superclasses.toMutableList()
    override val methods = MutableContainmentList(methods)
}

typealias PythonClassId = PythonDeclarationId<PythonClass>

@Serializable
data class PythonClassDocumentation(val description: String = "")
